File: LeetCode/slidingWindow/76.minimum-window-substring.py
# @before-stub-for-debug-begin
from python3problem76 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start
class Solution:
    def isVaild(self, charRequirement):
        # isVaild() => {return true if charCount's val all == 0}
        # print(charRequirement)
        for char in charRequirement:
            if charRequirement[char] > 0:
                return False
        return True
    def minWindow(self, s: str, t: str) -> str:
        # init optimums
        minLen = 1e9
        minLeftP = minRightP = 0

        leftP = rightP = 0

        # charRequirement
            #  {"charIn_t": "count of the char in substring"; init as required appearance}
        charRequirement = dict.fromkeys(t, 0)
        for char in t:
            charRequirement[char]+=1


        while (rightP < len(s)):
            # print ('start expand')
            curRChar = s[rightP]
            if (curRChar in charRequirement): 
                charRequirement[curRChar]-=1
            # print(s[leftP:rightP+1])
            isVaild = self.isVaild(charRequirement)
            # print('-----')

            while (isVaild):
                # print('start shrink')
                # Update global optimum
                if (rightP-leftP < minLen):
                    minLeftP = leftP
                    minRightP = rightP+1
                    minLen = rightP-leftP

                # shrink window
                curLChar = s[leftP]
                if curLChar in charRequirement:
                    charRequirement[curLChar]+=1
                leftP+=1 
                #re-eval vaildity 
                # print(s[leftP:rightP+1]) 
                isVaild = self.isVaild(charRequirement)
                # print('-----')
   
            
            rightP+=1
        # print(minLeftP)
        # print(minRightP)
        return s[minLeftP: minRightP]
        
# @lc code=end

# 在字符串 S 中使用双指针中的左右指针技巧，初始化 left = right = 0，把索引左闭右开区间 [left, right) 称为一个「窗口」。

# 先不断地增加 right 指针扩大窗口 [left, right)，直到窗口中的字符串符合要求（包含了 T 中的所有字符），此时更新最优的记录。: 相当于在寻找一个「可行解」

# 此时，我们停止增加 right，转而不断增加 left 指针缩小窗口 [left, right)，直到窗口中的字符串不再符合要求（不包含 T 中的所有字符了）。同时，每次增加 left，我们都要更新一轮最优的记录。

# 4、重复第 2 和第 3 步，左右指针轮流前进，窗口大小增增减减，窗口不断向右滑动, 直到 right 到达字符串 S 的尽头。

# O(n), 每一个字符最多进窗口一次然后出窗口一次

File: LeetCode_Go/List&LinkedList/76.minimum-window-substring.go
/*
 * @lc app=leetcode id=76 lang=golang
 *
 * [76] Minimum Window Substring
 */

// @lc code=start

// Given two strings `s` and `t`, return a substring in `s` such that contains all characters in `t`, and the length of the substring is minimized. If there is no such substring, return the empty string "". The testcases will be generated such that the answer is unique.

// Use hashmap to record remain times required for each character in `t`. Use a int to log number of characters currently satisfing. Run SlideWindow in `s`. 1.Expand right until all remaining time requirement is satisfied, then 2.shrink left until causing unsatisfications. Repeat the expansion and shrinking. Record the min satisfy length when shrinking.

// s = "ADOBECODEBANC", t = "ABC"
// ADOBEC => DOBEC(unsat) => DOBECODEBA => ODEBA(unsat) => ODEBANC => ANC(unsat) => <end>

func minWindow(s string, t string) string {
	// global answer
	var substring [2]int
	var minLength int = math.MaxInt

	// init hash map with number of appearance in t. goal: make all of them 0 or negative
	satisfyCount := make(map[byte]int)
	for i, _ := range t {
		satisfyCount[t[i]] += 1
	}

	// number of characters to be satisfied. goal: 0
	remainUnsat := len(satisfyCount)

	// slidewindow start
	left, right := 0, 0
	for right < len(s) {
		// expand right, if current char is one of t, update satisfyCount and remainUnsat
		if _, exists := satisfyCount[s[right]]; exists {
			satisfyCount[s[right]]--
			// this character just got satisfied
			if satisfyCount[s[right]] == 0 {
				remainUnsat--
			}
		}
		right++

		// when all satisfied, start log answer and start shrinking
		for remainUnsat == 0 {
			// log answer
			curLength := right - left
			if curLength < minLength {
				minLength = curLength
				substring[0], substring[1] = left, right-1
			}
			// remove a char, if the char is one of t, update satisfyCount and remainUnsat
			if _, exists := satisfyCount[s[left]]; exists {
				satisfyCount[s[left]]++
				if satisfyCount[s[left]] > 0 {
					remainUnsat++ // cause unsatisfiction
				}
			}
			left++
		}
	}

	if minLength == math.MaxInt {
		return ""
	}
	return s[substring[0] : substring[1]+1]
}

// @lc code=end

